src/v/coproc/service.cc

```diff
@@ -12,146 +12,162 @@
-          std::move(topics),
-          [this,
-           id](std::vector<enable_copros_request::data::topic_mode>& topics) {
-              return ssx::async_transform(
-                       topics.begin(),
-                       topics.end(),
-                       [this, id](enable_copros_request::data::topic_mode tm) {
-                           const auto v = enable_validator(tm.first, tm.second);
-                           if (v != erc::success) {
-                               return ss::make_ready_future<erc>(v);
-                           }
-                           return insert(
-                             id,
-                             model::topic_namespace(
-                               cluster::kafka_namespace, std::move(tm.first)),
-                             tm.second);
-                       })
-                .then([id](auto erc_vec) {
-                    return std::make_pair(id, std::move(erc_vec));
-                });
-          });
-    });
+    /// Then verify that the script id isn't already registered
+    return _pacemaker
+      .map_reduce0(
+        [id = input.id](pacemaker& pacemaker) {
+            /// ... to do this, query each shard for the id
+            return pacemaker.local_script_id_exists(id);
+        },
+        false,
+        std::logical_or<>())
+      .then([this, id = input.id, topics = std::move(input.topics)](
+              bool script_exists) mutable {
+          if (script_exists) {
+              /// Protocol dictates that if the script id already exists,
+              /// the topic acks reply will be a vector of 1 with the
+              /// script_id_dne code as the sole element
+              vlog(
+                coproclog.warn,
+                "Rejected attempt to register coprocessor with id {} as it "
+                "has "
+                "already been registered",
+                id);
+              return ss::make_ready_future<reply_type>(make_copro_error(
+                id, enable_response_code::script_id_already_exists));
+          }
+          return _pacemaker
+            .map([id, topics = std::move(topics)](pacemaker& pacemaker) {
+                /// Insert the script into the pacemaker, after ensuring a
+                /// double id insert is not possible
+                return pacemaker.add_source(id, topics);
+            })
+            .then([id](std::vector<std::vector<errc>> ack_codes) {
+                vassert(!ack_codes.empty(), "acks.size() must be > 0");
+                vassert(
+                  !ack_codes[0].empty(), "acks vector must contain values");
+                const bool all_equivalent = std::all_of(
+                  ack_codes.cbegin(),
+                  ack_codes.cend(),
+                  [s = ack_codes[0].size()](const std::vector<errc>& v) {
+                      return v.size() == s;
+                  });
+                vassert(all_equivalent, "Acks from all shards differ in size");
```

Comment:
Not really, because I am only asserting that a specific number of elements exist in each vector of vectors. Thats because there is some code right below this snippet that operates on indexes that exist within a bounds window. These assertions crash rp to prevent what would have later caused an out of bounds memory access.